client/snapshot.go

```diff
@@ -0,0 +1,151 @@
+package client
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/url"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/snapcore/snapd/snap"
+	"github.com/snapcore/snapd/strutil"
+)
+
+// errors on reading:
+var (
+	ErrSnapshotUnsupported  = errors.New("unsupported snapshot format")
+	ErrSnapshotSizeMismatch = errors.New("reported size and read size differ")
+	ErrSnapshotHashMismatch = errors.New("hash mismatch")
+)
+
+// errors about not finding stuff (should bubble out as something more user-friendly, probably)
+var (
+	ErrSnapshotNotFound      = errors.New("no snapshot by that ID")
+	ErrSnapshotSnapsNotFound = errors.New("the snapshot with the given ID does not contain the requested snaps")
+)
+
+// A SnapshotOp is used to request an operation on a snapshot
+type SnapshotOp struct {
+	ID     uint64   `json:"ID"`
+	Action string   `json:"action"`
+	Snaps  []string `json:"snaps,omitempty"`
+	Homes  []string `json:"snaps,omitempty"`
+}
+
+func (op *SnapshotOp) String() string {
+	// verb of snapshot #N [for snaps %q] [for homes %q]
+	var snaps string
+	var homes string
+	if len(op.Snaps) > 0 {
+		snaps = " for snaps " + strutil.Quoted(op.Snaps)
+	}
+	if len(op.Homes) > 0 {
+		homes = " for homes " + strutil.Quoted(op.Homes)
+	}
+	return fmt.Sprintf("%s of snapshot #%d%s%s", strings.Title(op.Action), op.ID, snaps, homes)
+}
+
+// A Snapshot is a collection of archives with a simple metadata json file
+// (and hashsums of everything)
+//
+// XXX can snapd sign these?
+type Snapshot struct {
+	// ID of the snapshot group (a snapshot group is a "snap save" invocation)
+	ID uint64 `json:"snapshot"`
+	// the snap this data is for
+	Snap string `json:"snap"`
+	// the snap's revision
+	Revision snap.Revision `json:"revision"`
+	// the snap's version (optional)
```

Comment:
Being optional reads a bit strange here. Why we'd not copy over the version when doing a snapshot?  If that's not that it means, the comment should probably be dropped as that's how it sounds.

Reply:
It's optional because we can have snaps installed that don't have a version at all, and might still want/need to snapshot them.
```console
$ mkdir -p asnap/meta
$ echo '{name: asnap, apps: {asnap: {command: ../../../bin/echo}}}' > asnap/meta/snap.yaml
$ snap try asnap
asnap  mounted from .../asnap
$ asnap hello
hello
$ snap list asnap
Name   Version  Rev  Developer  Notes
asnap           x1              try
```

Reply:
A bit surprising to see that version is not a required field. I think we should fix that, assuming that snapcraft or the store are already enforcing that uniqueness.

Reply:
[done](https://forum.snapcraft.io/t/snapd-enforcing-snap-versions/3974) (and now fixing this...)